explorer/interpreter/matching_impl_set.cpp

```diff
@@ -0,0 +1,163 @@
+
+ private:
+  // Most kinds of value don't contribute to the signature.
+  void VisitValue(const Value*) {}
+
+  void VisitValue(const TupleType* tuple) {
+    for (auto* elem : tuple->elements()) {
+      Collect(elem);
+    }
+  }
+
+  void VisitValue(const TypeType*) { Collect(ValueKey::TypeType); }
+
+  void VisitValue(const BoolType*) { Collect(ValueKey::BoolType); }
+
+  void VisitValue(const IntType*) { Collect(ValueKey::IntType); }
+
+  void VisitValue(const StringType*) { Collect(ValueKey::StringType); }
+
+  void VisitValue(const StaticArrayType* array) {
+    Collect(ValueKey::ArrayType);
+    Collect(&array->element_type());
+  }
+
+  void VisitValue(const PointerType* pointer) {
+    Collect(ValueKey::PointerType);
+    Collect(&pointer->pointee_type());
+  }
+
+  void VisitValue(const StructType* struct_type) {
+    for (auto [name, type] : struct_type->fields()) {
+      Collect(type);
+    }
+  }
+
+  void VisitValue(const NominalClassType* class_type) {
+    VisitDeclarationAndArgs(class_type->declaration(), class_type->bindings());
+  }
+
+  void VisitValue(const MixinPseudoType* mixin_type) {
+    VisitDeclarationAndArgs(mixin_type->declaration(), mixin_type->bindings());
+  }
+
+  void VisitValue(const InterfaceType* iface_type) {
+    VisitDeclarationAndArgs(iface_type->declaration(), iface_type->bindings());
+  }
+
+  void VisitValue(const NamedConstraintType* constraint_type) {
+    VisitDeclarationAndArgs(constraint_type->declaration(),
+                            constraint_type->bindings());
+  }
+
+  void VisitValue(const ChoiceType* choice_type) {
+    VisitDeclarationAndArgs(choice_type->declaration(),
+                            choice_type->bindings());
+  }
+
+  void VisitDeclarationAndArgs(const Declaration& declaration,
+                               const Bindings& bindings) {
+    Collect(match_->parent_->GetKeyForDeclaration(declaration));
+    for (auto [key, value] : bindings.args()) {
+      Collect(value);
+    }
+  }
+
+ private:
+  Match* match_;
+};
+
+auto MatchingImplSet::GetKeyForDeclaration(const Declaration& declaration)
+    -> ValueKey {
+  auto [it, added] = declaration_keys_.insert(
+      {&declaration, ValueKey(int(ValueKey::FirstDeclarationKey) +
+                              declaration_keys_.size())});
+  return it->second;
+}
+
+MatchingImplSet::Match::Match(Nonnull<MatchingImplSet*> parent,
+                              Nonnull<const ImplScope::Impl*> impl,
+                              Nonnull<const Value*> type,
+                              Nonnull<const Value*> interface)
+    : parent_(parent), impl_(impl), type_(type), interface_(interface) {
+  // Build our signature.
+  LeafCollector collector(this);
+  collector.Collect(type);
+  collector.Collect(interface);
+
+  parent_->matches_.push_back(this);
+}
+
+MatchingImplSet::Match::~Match() {
+  CARBON_CHECK(parent_->matches_.back() == this) << "match stack broken";
+  parent_->matches_.pop_back();
```

Comment:
```suggestion
  Match* removed = parent_->matches_.pop_back_val();
  CARBON_CHECK(removed == this) << "match stack broken";
```

I assume you want to keep the action outside the CHECK, but I'd lean towards this in order to only look at the back once.

Reply:
`std::vector` doesn't have a `pop_back_val`; that's an LLVM `SmallVector`-specific mechanism.

Reply:
Oops, and we aren't using SmallVector here...